components/server/src/authorization/perms.ts

```diff
@@ -31,3 +42,66 @@ export type OrganizationOperation =
     | "org_authprovider_write"
     // Ability to read Organization Auth Providers.
     | "org_authprovider_read";
+
+export type ResourceType = "organization";
+
+export type SubjectType = "user";
+
+export type CheckResult = {
+    permitted: boolean;
+    err?: Error;
+    response?: v1.CheckPermissionResponse;
+};
+
+export const NotPermitted = { permitted: false };
+
+export const PermissionChecker = Symbol("PermissionChecker");
+
+export interface PermissionChecker {
```

Comment:
So the `guard` layer of APIs should only use a `PermissionChecker`. Once that check passes, and we create a resource, we'll want to use a `PermissionGranter`. They can both be implemented by the same class (`Authorizer`), but the type signatures should be constrained. The goal is to use composition, rather than exposing everything. This makes it easier to actually write stub implementations for tests (if we had any..).

The type does in fact rely on the `spicedb` type. Not ideal, but at this stage I wanted to reduce re-definition of types. If we needed to abstract, it would be relatively easy to swap the argument type for a custom one.

Reply:
Thank you, makes sense 👍